nixos/modules/programs/rust-motd.nix

```diff
@@ -5,6 +5,34 @@ with lib;
 let
   cfg = config.programs.rust-motd;
   format = pkgs.formats.toml { };
+
+  orderedSections = listToAttrs
+    (imap0
+      (i: attr: nameValuePair "env${toString i}" {
+        ${attr} = cfg.settings.${attr};
+      })
+      cfg.order);
+
+  # Order the sections in the TOML according to the order of sections
+  # in `cfg.order`.
+  # This is done by
+  # * creating an attribute set with keys `env0`/`env1`/.../`envN`
+  #   where `env0` contains the first section and `envN` the last
+  #   (in the form of `{ sectionName = { /* ... */ }}`)
+  # * the declarations of `env0` to `envN` in ascending order are
+  #   concatenated with `jq`. Now we have a JSON representation of
+  #   the config in the correct order.
+  # * this is piped to `json2toml` to get the correct format for rust-motd.
+  motdConf = pkgs.runCommand "motd.conf"
+    (orderedSections // {
+      __structuredAttrs = true;
+      nativeBuildInputs = [ pkgs.remarshal pkgs.jq ];
+    })
+    ''
+      cat "$NIX_ATTRS_JSON_FILE" \
+        | jq '${concatMapStringsSep " + " (key: ''."${key}"'') (attrNames orderedSections)}' \
+        | json2toml /dev/stdin "$out"
+    '';
```

Comment:
It's generally brittle to interpolate strings, escaping can get messed up. Instead I recommend an approach like this:

```nix
  motdConf = pkgs.runCommand "motd.conf"
    {
      __structuredAttrs = true;
      inherit (cfg) settings order;
      nativeBuildInputs = [ pkgs.remarshal pkgs.jq ];
    }
    ''
      jq '.settings as $settings
          | .order
          | map({ key: "\(.)", value: $settings."\(.)" })
          | from_entries' "$NIX_ATTRS_JSON_FILE" \
        | json2toml /dev/stdin "$out"
    '';
```

No string interpolation -> no escaping problems :)